omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,171 @@
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+/**
+ * @brief The reordering namespace.
+ *
+ * @ingroup reorder
+ */
+namespace rcm {
+
+
+template <typename ValueType, typename IndexType>
+void get_degree_of_nodes(
+    std::shared_ptr<const OmpExecutor> exec,
+    std::shared_ptr<matrix::SparsityCsr<ValueType, IndexType>> adjacency_matrix,
+    std::shared_ptr<gko::Array<IndexType>> node_degrees)
+{
+    auto num_rows = adjacency_matrix->get_size()[0];
+    auto adj_ptrs = adjacency_matrix->get_row_ptrs();
+    auto node_deg = node_degrees->get_data();
+
+#pragma omp parallel for
+    for (auto i = 0; i < num_rows; ++i) {
+        node_deg[i] = adj_ptrs[i + 1] - adj_ptrs[i];
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_RCM_GET_DEGREE_OF_NODES_KERNEL);
+
+
+template <typename IndexType>
+IndexType find_index(std::vector<std::pair<IndexType, IndexType>> &a,
+                     IndexType x)
+{
+    for (auto i = 0; i < a.size(); i++)
+        if (a[i].first == x) return i;
+    return -1;
+}
+
+
+template <typename ValueType, typename IndexType>
+void get_permutation(
+    std::shared_ptr<const OmpExecutor> exec, size_type num_vertices,
+    std::shared_ptr<matrix::SparsityCsr<ValueType, IndexType>> adjacency_matrix,
+    std::shared_ptr<Array<IndexType>> node_degrees,
+    std::shared_ptr<matrix::Permutation<IndexType>> permutation_mat,
+    std::shared_ptr<matrix::Permutation<IndexType>> inv_permutation_mat)
+{
+    IndexType num_vtxs = static_cast<IndexType>(num_vertices);
+    auto adj_ptrs = adjacency_matrix->get_row_ptrs();
+    auto adj_idxs = adjacency_matrix->get_col_idxs();
+    auto node_deg = node_degrees->get_data();
+    auto permutation_arr = permutation_mat->get_permutation();
+    auto inv_permutation_arr = inv_permutation_mat->get_permutation();
+
+    std::queue<IndexType> q;
+    std::vector<IndexType> r;
+    std::vector<std::pair<IndexType, IndexType>> not_visited;
+
+    for (auto i = 0; i < num_vtxs; ++i) {
+        not_visited.push_back(std::make_pair(i, node_deg[i]));
+    }
+
+    while (not_visited.size()) {
+        // choose this better.
+        IndexType min_node_index = 0;
+
+        for (auto i = 0; i < not_visited.size(); i++) {
+            if (not_visited[i].second < not_visited[min_node_index].second) {
+                min_node_index = i;
+            }
+        }
+        q.push(not_visited[min_node_index].first);
+
+        not_visited.erase(
+            not_visited.begin() +
+            find_index(not_visited, not_visited[q.front()].first));
+
+        // Simple BFS
+        while (!q.empty()) {
+            std::vector<IndexType> to_sort;
+
+            for (IndexType i = 0; i < num_vtxs; i++) {
+                if (i != q.front() && find_index(not_visited, i) != -1) {
```

Comment:
If I see this correctly, this is the section: "Get all Nodes adjacent to the currently selected one". But isn't this collecting simply **all** nodes still left? Maybe you forgot addding the condition " ... and adjacent to `q.front()`"?

Reply:
Yep, I did forget that. Please feel free to scrap this (and the reference) kernel and rewrite the whole thing. This was something that I did not put any thought into, so the it is probably not a good point to start off from.